src/com/seafile/seadroid2/data/DataManager.java

```diff
@@ -122,84 +112,23 @@ public static File getTempFile(String path, String oid) {
         return new File(getExternalCacheDirectory() + "/" + oid);
     }
     
-    public static File getAvatarCacheDirectory() {
-        return new File(getExternalCacheDirectory() + "/avatar");
+    public static File getThumbnailCacheDirectory() {
+        return new File(getExternalCacheDirectory() + "/thumbnails");
     }
 
-    public void calculateThumbnail(String repoName, String repoID, String path, String oid) {
-        final int THUMBNAIL_SIZE = caculateThumbnailSizeOfDevice();
-        FileOutputStream out = null;
-        try {
-            File file = getLocalRepoFile(repoName, repoID, path);
-            if (!file.exists())
-                return;
-            // if (file.length() > MAX_GEN_CACHE_THUMB)
-            //     return;
-
-            Bitmap imageBitmap = BitmapUtil.calculateThumbnail(file.getPath(), THUMBNAIL_SIZE);
-            if (imageBitmap == null) {
-                return;
-            }
-            ByteArrayOutputStream baos = new ByteArrayOutputStream();
-            imageBitmap.compress(Bitmap.CompressFormat.PNG, 100, baos);
-            byte[] byteArray = baos.toByteArray();
-            File thumb = getThumbFile(oid);
-            out = new FileOutputStream(thumb);
-            out.write(byteArray);
-        } catch (IOException e) {
-            Log.i(DEBUG_TAG, "Failed to write thumbnail : " + e.getMessage());
-        } finally {
-            try {
-                if(out != null) {
-                    out.close();
-                }
-            }
-            catch(IOException ioe) {
-            }
-        }
-    }
+    public String getThumbnailLink(String repoName, String repoID, String filePath, int size) {
+        File file = getLocalRepoFile(repoName, repoID, filePath);
 
-    /**
-     * Caculate the thumbnail of an image directly when its size is less than {@link #MAX_DIRECT_SHOW_THUMB}
-     */
-    public Bitmap getThumbnail(File file) {
-        try {
-            if (!file.exists())
-                return null;
-
-            final int THUMBNAIL_SIZE = caculateThumbnailSizeOfDevice();
-
-            return BitmapUtil.calculateThumbnail(file.getPath(), THUMBNAIL_SIZE);
-        } catch (Exception ex) {
-            return null;
-        }
-    }
-
-
-    public static int caculateThumbnailSizeOfDevice() {
-        DisplayMetrics metrics = SeadroidApplication.getAppContext().getResources().getDisplayMetrics();
-
-        switch(metrics.densityDpi) {
-        case DisplayMetrics.DENSITY_LOW:
-            return 36;
-        case DisplayMetrics.DENSITY_MEDIUM:
-            return 48;
-        case DisplayMetrics.DENSITY_HIGH:
-            return 72;
-        case DisplayMetrics.DENSITY_XHIGH:
-            return 96;
-        default:
-            return 36;
+        // use locally cached file if available
+        if (file.exists()) {
+            return "file://" + file.getAbsolutePath();
+        } else {
+            return sc.getThumbnailLink(repoID, filePath, size);
```

Comment:
`sc.getThumbnailLink` and `encodeThumbnailFilePath` are not network requests, so put them in `DataManager`.